ambari-server/src/main/java/org/apache/ambari/server/events/listeners/upgrade/MpackInstallStateListener.java

```diff
@@ -159,10 +176,150 @@ public void onMpackEvent(MpackRegisteredEvent event) {
+      LOG.error(
+          "Unable to parse the installation structured output for command {} for {} on host {}",
+          commandReport.getTaskId(), event.getRole(), hostName, jsonException);
+    }
+
+    if (null == structuredOutput || null == structuredOutput.mpackId) {
+      LOG.error("The structured output for command {} for {} on {} did not contain an mpack ID",
+          commandReport.getTaskId(), event.getRole(), hostName);
+    } else {
+      mpackId = structuredOutput.mpackId;
+    }
+
+    if (!StringUtils.equals(HostRoleStatus.COMPLETED.name(), commandReport.getStatus())) {
+      LOG.warn(
+          "Command {} for {} did not complete on {}. The management pack installation state will be updated to {}.",
+          commandReport.getTaskId(), event.getRole(), hostName, mpackInstallState);
+    } else {
+      mpackInstallState = MpackInstallState.INSTALLED;
+    }
+
+    Lock lock = m_locksByHost.get(hostName);
+    lock.lock();
+    try {
+      // if no mpack ID, then find all INSTALLING and set them to failed
+      if (null == mpackId) {
+        List<MpackHostStateEntity> mpackHostStateEntities = mpackHostStateDAO.findByHostAndInstallState(
+            hostName, MpackInstallState.INSTALLING);
+
+        for (MpackHostStateEntity mpackHostStateEntity : mpackHostStateEntities) {
+          mpackHostStateEntity.setState(MpackInstallState.INSTALL_FAILED);
+          mpackHostStateDAO.merge(mpackHostStateEntity);
+        }
+      } else {
+        // find the mpack for the host and update the state
+        MpackHostStateEntity mpackHostStateEntity = mpackHostStateDAO.findByMpackAndHost(
+            mpackId, hostName);
+
+        // an odd case, but we'll plan for it
+        if( null == mpackHostStateEntity ) {
+          HostDAO hostDAO = m_hostDAOProvider.get();
+          MpackDAO mpackDAO = m_mpackDAOProvider.get();
+
+          MpackEntity mpackEntity = mpackDAO.findById(mpackId);
+          HostEntity hostEntity = hostDAO.findByName(hostName);
+
+          mpackHostStateEntity = new MpackHostStateEntity(mpackEntity, hostEntity, mpackInstallState);
+          mpackHostStateDAO.create(mpackHostStateEntity);
+
+          hostEntity.getMpackInstallStates().add(mpackHostStateEntity);
+          hostEntity = hostDAO.merge(hostEntity);
+        } else {
+          // don't set it if it's already marked as installed
+          if (mpackHostStateEntity.getState() != MpackInstallState.INSTALLED) {
+            mpackHostStateEntity.setState(mpackInstallState);
+            mpackHostStateEntity = mpackHostStateDAO.merge(mpackHostStateEntity);
+          }
+        }
+      }
+    } catch (Throwable throwable) {
+      LOG.error(throwable.getMessage(), throwable);
+    } finally {
+      lock.unlock();
+    }
+  }
+
+  /**
+   * Used only to parse the structured output when installing mpacks.
+   */
+  private static class InstallCommandStructuredOutput {
+    /**
+     * Either SUCCESS or FAIL
```

Comment:
can it be IN_PROGRESS or something like that? Would agent send FAIL after sudden agent restart?

Reply:
No, actually - I don't even think this field is going to be used. The `CommandReportReceivedEvent` is only sent when the command is in a completed state:

```
      if (HostRoleStatus.valueOf(report.getStatus()).isCompletedState()) {
        CommandReportReceivedEvent event = new CommandReportReceivedEvent(
            clusterId, hostname, report);

        commandReportEventPublisher.publish(event);
      }
```